src/main/java/org/opensearch/neuralsearch/search/collector/CollapseDocSourceGroupSelector.java

```diff
@@ -0,0 +1,255 @@
+            if (values.advanceExact(doc)) {
+                hasValue = true;
+                value = values.longValue();
+                return State.ACCEPT;
+            } else {
+                hasValue = false;
+                return State.SKIP;
+            }
+        }
+
+        @Override
+        public Long currentValue() {
+            return hasValue ? value : null;
+        }
+
+        @Override
+        public Long copyValue() {
+            return currentValue();
+        }
+
+        @Override
+        public void setNextReader(LeafReaderContext readerContext) throws IOException {
+            LeafReader reader = readerContext.reader();
+            DocValuesType type = getDocValuesType(reader, field);
+            if (type == null || type == DocValuesType.NONE) {
+                values = DocValues.emptyNumeric();
+                return;
+            }
+            switch (type) {
+                case NUMERIC:
+                    values = DocValues.getNumeric(reader, field);
+                    break;
+
+                case SORTED_NUMERIC:
+                    final SortedNumericDocValues sorted = DocValues.getSortedNumeric(reader, field);
+                    values = DocValues.unwrapSingleton(sorted);
+                    if (values == null) {
+                        values = new AbstractNumericDocValues() {
+
+                            private long value;
+
+                            @Override
+                            public boolean advanceExact(int target) throws IOException {
+                                if (sorted.advanceExact(target)) {
+                                    if (sorted.docValueCount() > 1) {
+                                        throw new IllegalStateException(
+                                            "failed to collapse " + target + ", the collapse field must be single valued"
+                                        );
+                                    }
+                                    value = sorted.nextValue();
+                                    return true;
+                                } else {
+                                    return false;
+                                }
+                            }
+
+                            @Override
+                            public int docID() {
+                                return sorted.docID();
+                            }
+
+                            @Override
+                            public long longValue() throws IOException {
+                                return value;
+                            }
+
+                        };
+                    }
+                    break;
+
+                default:
+                    throw new IllegalStateException("unexpected doc values type " + type + "` for field `" + field + "`");
+            }
+        }
+
+        @Override
+        public void setScorer(Scorable scorer) throws IOException {}
```

Comment:
to confirm my understanding - this means we don't care about scores, just doc ids are needed?

Reply:
why exactly usage of GroupSelector is out of the picture?

Reply:
We are using a GroupSelector, just grouping based on a collapse value.